Fixes #32486 - Fedora 34 / RHEL 9 - Ignore Anaconda kernel boot parameters without 'inst.' prefix

ShuffleBox opened the pull request:
Fixes [#32486](https://projects.theforeman.org/issues/32486)

Summary: Fedora 34 & RHEL 9 have removed support for deprecated kernel install parameters.  This breaks install with default kickstart templates.
This pull request adds some If/Else logic to provide the expected presentation of the kernel parameters.

An additional item that needs thought/review should this PR be accepted is if it applies to other templates that I do not have experience with from my environment.

lzap commented on app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb:
```diff
@@ -29,7 +29,12 @@ snippet: true
+    options.push("inst.ks.sendmac")
+  else
+    options.push("kssendmac", "ks.sendmac", "inst.ks.sendmac")  
+  end
```
Thank you, I am gonna ask for more work but if you don't feel comfortable just say a word and we can merge this as-is.

I think we can move the "ks/inst.ks" option in here, since kickstart_kernel_options snippet is used pretty much in all kickstart types, that would be a nice place to have it. What you think?

ShuffleBox replied:
I think it seems like a good way to DRY, which is usually a plus.  I believe I moved the change into the kickstart_kernel_options snippet.

lzap commented on app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb:
```diff
@@ -29,7 +29,12 @@ snippet: true
   # both current and legacy syntax provided
-  options.push("kssendmac", "ks.sendmac", "inst.ks.sendmac")
+
+  if (is_fedora && os_major >= 34) || (rhel_compatible && os_major >= 9)
```
Please remove the extra line, the comment above does belong to lines 33-37 actually.

ShuffleBox replied:
The extra line should now be removed.

lzap reviewed (changes requested):
The changes look clean, I have one comment on an extra empty line and one suggestion - if you could actually move the ks to the common snippet and remove it from all the templates where it is referencing (just grep the codebase) that would be great.

Finally, this will fail to pass our template rendering tests, what you need to do is either edit them manually:

```
$ ag ks=http test/unit/foreman/renderer/snapshots
test/unit/foreman/renderer/snapshots/ProvisioningTemplate/PXEGrub/Kickstart default PXEGrub.snap.txt
8:  kernel (nd)/../ ks=http://foreman.some.host.fqdn/unattended/provision  BOOTIF=01-00-f0-54-1a-7e-e0 kssendmac ks.sendmac inst.ks.sendmac fips=1

test/unit/foreman/renderer/snapshots/ProvisioningTemplate/PXEGrub2/Kickstart default PXEGrub2.snap.txt
7:  linuxefi  ks=http://foreman.some.host.fqdn/unattended/provision  BOOTIF=01-00-f0-54-1a-7e-e0 kssendmac ks.sendmac inst.ks.sendmac fips=1

test/unit/foreman/renderer/snapshots/ProvisioningTemplate/PXELinux/Kickstart default PXELinux.snap.txt
10:APPEND initrd= ks=http://foreman.some.host.fqdn/unattended/provision  BOOTIF=01-00-f0-54-1a-7e-e0 kssendmac ks.sendmac inst.ks.sendmac fips=1
```

or use the following command to regenerate them automatically:

    RAILS_ENV=test bundle exec rake snapshots:generate

ekohl reviewed (changes requested):
https://docs.fedoraproject.org/ar/fedora/f33/release-notes/sysadmin/Installation/ states:
> Anaconda now displays a warning if you use an Anaconda-specific boot option without the inst. prefix. In the next release (Fedora 34), support for using these options without a prefix will be removed completely.

So at least Fedora 33 deprecated it. We can also consider finding the earliest version that does support it but it should at least be `>= 33`.

ShuffleBox reviewed (commented):
I believe I've updated the PR as requested except for updating the tests.  I'm going to take few to try to sort it.

ShuffleBox reviewed (commented):
Changed version check to Fedora 33 as requested by ekohl.

ShuffleBox commented:
> The changes look clean, I have one comment on an extra empty line and one suggestion - if you could actually move the ks to the common snippet and remove it from all the templates where it is referencing (just grep the codebase) that would be great.
> 
> Finally, this will fail to pass our template rendering tests, what you need to do is either edit them manually:
> 
> ```
> $ ag ks=http test/unit/foreman/renderer/snapshots
> test/unit/foreman/renderer/snapshots/ProvisioningTemplate/PXEGrub/Kickstart default PXEGrub.snap.txt
> 8:  kernel (nd)/../ ks=http://foreman.some.host.fqdn/unattended/provision  BOOTIF=01-00-f0-54-1a-7e-e0 kssendmac ks.sendmac inst.ks.sendmac fips=1
> 
> test/unit/foreman/renderer/snapshots/ProvisioningTemplate/PXEGrub2/Kickstart default PXEGrub2.snap.txt
> 7:  linuxefi  ks=http://foreman.some.host.fqdn/unattended/provision  BOOTIF=01-00-f0-54-1a-7e-e0 kssendmac ks.sendmac inst.ks.sendmac fips=1
> 
> test/unit/foreman/renderer/snapshots/ProvisioningTemplate/PXELinux/Kickstart default PXELinux.snap.txt
> 10:APPEND initrd= ks=http://foreman.some.host.fqdn/unattended/provision  BOOTIF=01-00-f0-54-1a-7e-e0 kssendmac ks.sendmac inst.ks.sendmac fips=1
> ```
> 
> or use the following command to regenerate them automatically:
> 
> ```
> RAILS_ENV=test bundle exec rake snapshots:generate
> ```
I've been working to setup a functioning development environment to use snapshots:generate, but am struggling a bit to get it running properly.

Could I have a bit of guidance on manually modifying the tests, as I think that'll get this resolved faster?  
Are the tests looking for 'inst.ks=' vs 'ks=', or something else?

ShuffleBox reviewed (commented):
I got my dev environment to regenerate the snapshots.  However, I'm not sure if the results are what was originally expected.

lzap commented on app/views/unattended/provisioning_templates/iPXE/kickstart_default_ipxe.erb:
```diff
@@ -27,8 +27,8 @@ ping --count 1 ${netX/gateway} || echo Ping to Gateway failed or ping command no
 imgstat
 sleep 2
-boot
+boot
```
Ah can you revert this one, here is special iPXE handling for ks so please just do the normal `ks` to `inst.ks` change here. Also keep the trailing newline that can cause issues sometimes (not here specifically tho).

ShuffleBox replied:
These should both be fixed in e0197b0.  Thank you for your patience while I stumble through some of this.  It's been a multifaceted learning experience.

lzap reviewed (dismissed):
Yes it now looks great, would you mind correcting the iPXE special case? It is a bit different there, other than that we can trigger tests and merge, unless @ekohl has other concerns. Thank you for the refactoring! This is very much appreciated.

Edit: I still request one change, I just accidentally clicked on APPROVE :-)

lzap commented on app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb:
```diff
@@ -29,7 +29,11 @@ snippet: true
+  if (is_fedora && os_major >= 33) || (rhel_compatible && os_major >= 9)
+    options.push("inst.ks=#{foreman_url('provision')}", "inst.ks.sendmac")
+  else
+    options.push("ks=#{foreman_url('provision')}", "kssendmac", "ks.sendmac", "inst.ks.sendmac")  
```
Nitpick: When rebasing I think you can drop `inst.ks.sendmac` from the else case, we were sending it "just for case" but obviously the ks.sendmac should work up until F33.

lzap commented:
Looks great can you squash into a single commit, rebase, regenerate snapshots (we merged a big PR last week which caused conflicts) and we can merge!

ShuffleBox commented:
> Looks great can you squash into a single commit, rebase, regenerate snapshots (we merged a big PR last week which caused conflicts) and we can merge!

I believe I've squashed and rebased successfully.  Some extra scrutiny would be warranted and appreciated, but hopefully it's good to go.

lzap commented:
I am so sorry, before I was able to get back to the PR, we merged complete rename of all snapshots, all whitespace is now underscores. So would you mind (for the last time) do:

1. rebase (you will get conflicts just drop all changes in the snapshot dir prior rebasing to avoid them)
2. `rm -rf test/unit/foreman/renderer/snapshots/*` (you really need to delete all of them)
3. `RAILS_ENV=test bundle exec rake snapshots:generate`
4. make sure you have a single commit, amend

Then we are good to go!

ShuffleBox committed:
Fixes #32486 - Fedora34/RHEL9 deprecated kernel boot parameters

ShuffleBox commented:
Rebasing from upstream is still a bit awkward for me, but I think I got it!  Please let me know if there's anything I need to take another look/try at.

lzap commented:
I tested locally and it works fine, thanks. Snapshots do match. RTM

lzap commented:
[test foreman]

lzap commented on test/unit/foreman/renderer/snapshots/ProvisioningTemplate/PXEGrub/Kickstart_default_PXEGrub.host4and6dhcp.snap.txt:
```diff
@@ -5,7 +5,7 @@ timeout=10
 title Kickstart default PXEGrub
   root (nd)
-  kernel (nd)/../boot/centos-mirror-nrm0GtSX1ZC5-vmlinuz ks=http://foreman.some.host.fqdn/unattended/provision  network ksdevice=bootif ks.device=bootif BOOTIF=01-00-f0-54-1a-7e-e0 kssendmac ks.sendmac inst.ks.sendmac ip=dhcp nameserver=192.168.42.2 nameserver=192.168.42.3 fips=1
+  kernel (nd)/../boot/centos-mirror-nrm0GtSX1ZC5-vmlinuz  network ksdevice=bootif ks.device=bootif BOOTIF=01-00-f0-54-1a-7e-e0 ks=http://foreman.some.host.fqdn/unattended/provision kssendmac ks.sendmac ip=dhcp nameserver=192.168.42.2 nameserver=192.168.42.3 fips=1
```
The extra space before `network` was already there, let's not make this any longer.

lzap reviewed (approved):
Thank you for your patience, I am drowning in github notifications...

Thanks for the patch, this is RTM.

lzap commented:
There are exactly zero snapshot tests for RHEL9 at the moment, this is passing just fine. Thanks.

lzap commented:
> So at least Fedora 33 deprecated it. We can also consider finding the earliest version that does support it but it should at least be `>= 33`.

@ekohl you good?

ekohl commented on app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb:
```diff
@@ -29,7 +29,11 @@ snippet: true
+  if (is_fedora && os_major >= 33) || (rhel_compatible && os_major >= 9)
+    options.push("inst.ks=#{foreman_url('provision')}", "inst.ks.sendmac")
+  else
+    options.push("ks=#{foreman_url('provision')}", "kssendmac", "ks.sendmac")  
```
I'm not sure if I would remove `inst.ks.sendmac` here, given that it was present previously. Keeping it in feels safer, but no objection if you think it's safe.

lzap replied:
The correct (and required) argument is `inst.ks.sendmac` both of the other forms were deprecated. Thanks.
